### PR TITLE
HPCC-31882 Reduce the output from thor slaves

### PR DESCRIPTION
**Pull request by richardkchapman**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: thorlcr/activities/hashdistrib/thhashdistribslave.cpp [hunk 1]
```diff
@@ -879,20 +879,22 @@ class CDistributorBase : implements IHashDistributor, implements IExceptionHandl
                     unsigned dest=iter->get();
                     CTarget *target = targets.item(dest);
                     Owned<CSendBucket> bucket = target->getBucketClear();
-                    HDSendPrintLog3("Looking at last bucket: target=%s, size=%u", target->queryInfo(), bucket.get()?bucket->querySize():-1);
                     if (bucket && bucket->querySize())
                     {
-                        HDSendPrintLog3("Sending last bucket(s): target=%s, size=%u", target->queryInfo(), bucket->querySize());
+                        //HDSendPrintLog3("Sending last bucket(s): target=%s, size=%u", target->queryInfo(), bucket->querySize());
```
Comment (jakesmith) [inline thorlcr/activities/hashdistrib/thhashdistribslave.cpp:884]: not output in release anyway, but yes too verbose in debug build.
I'd be content with these being removed altogether, they haven't been useful in a long time.

File: thorlcr/graph/thgraph.hpp [hunk 2]
```diff
@@ -1134,6 +1136,8 @@ class graph_decl CActivityBase : implements CInterfaceOf<IThorRowInterfaces>, im
     inline void setInitialized(bool tf) { initialized = tf; }
     inline bool queryTimeActivities() const { return timeActivities; }
     inline roxiemem::RoxieHeapFlags queryHeapFlags() const { return defaultRoxieMemHeapFlags; }
+    inline bool traceStartStop() const { return false; }
+    inline bool traceActivity() const { return doTrace(traceThorActivities); }  // Could support overriding per activity via hint or other means, in the future
```
Comment (jakesmith) [inline thorlcr/graph/thgraph.hpp:1140]: there is an existing mechanism/function for per activity logging : container.doLogging()
However, at the moment it is:
a) always on for all graphs except child graphs
b) configurable per graph, not per activity (but would be easy to change so individual acts could be activated instead) :
```
void CGraphBase::init()
{
    bool log = queryJob().queryForceLogging(queryGraphId(), (NULL == queryOwner()) || isGlobal());
    setLogging(log);
}
```

This mechanism will still be active when traceThorActivities is enabled.

File: system/jlib/jtrace.hpp [hunk 3]
```diff
@@ -383,6 +385,23 @@ constexpr std::initializer_list<TraceOption> roxieTraceOptions
...
+    TRACEOPT(traceDetailed),
+    TRACEOPT(traceMax),
+    TRACEOPT(traceHttp),
+    TRACEOPT(traceSockets),
+    TRACEOPT(traceCassandra),
+    TRACEOPT(traceMongoDB),
+    TRACEOPT(traceCouchbase),
+    TRACEOPT(traceFilters),
+    TRACEOPT(traceKafka),
+    TRACEOPT(traceJava),
```
Comment (jakesmith) [inline system/jlib/jtrace.hpp:401]: looks like a lot of these should be removed/are not going to be used by Thor (cassandra/mongodb/counchbase ..)

File: thorlcr/activities/msort/thmsortslave.cpp [hunk 4]
```diff
@@ -156,10 +160,14 @@ class MSortSlaveActivity : public CSlaveActivity
...
         {
-            ActPrintLog("SORT waiting barrier.2");
+            if (traceActivity())
+                ActPrintLog("SORT waiting barrier.2");
             barrier->wait(false);
-            ActPrintLog("SORT barrier.2 raised");
-            ActPrintLog("SORT waiting for merge");
+            if (traceActivity())
+            {
+                ActPrintLog("SORT barrier.2 raised");
```
Comment (jakesmith) [inline thorlcr/activities/msort/thmsortslave.cpp:168]: these types of messages have often been useful in identifying where and which workers are stalled.. probably not the best way, but will probably make it harder to diagnose which/where without them.
Maybe we should be logging, but not to logging backend, but only to postmortem log.

File: thorlcr/graph/thgraph.cpp [hunk 5]
```diff
@@ -2462,13 +2461,17 @@ class CGraphExecutor : implements IGraphExecutor, public CInterface
         }
         return NULL;
     }
+
+    inline bool traceExecution() const { return false; }
```
Comment (jakesmith) [inline thorlcr/graph/thgraph.cpp:2465]: needs a 'traceThorExecution' flag and a doTrace call?

File: thorlcr/graph/thgraph.hpp [hunk 6]
```diff
@@ -883,6 +883,7 @@ class graph_decl CJobBase : public CInterfaceOf<IInterface>, implements IExcepti
     memsize_t keyNodeCacheBytes = 0;
     memsize_t keyLeafCacheBytes = 0;
     memsize_t keyBlobCacheBytes = 0;
+    TraceFlags oldTraceFlags;
```
Comment (jakesmith) [inline thorlcr/graph/thgraph.hpp:886]:
```
TraceFlags oldTraceFlags = TraceFlags::None
```
?
I know gets initialize in an early method (but not ctor), but I am a fan of member initializing everything possible,

File: thorlcr/graph/thgraph.hpp [hunk 7]
```diff
@@ -1134,6 +1136,8 @@ class graph_decl CActivityBase : implements CInterfaceOf<IThorRowInterfaces>, im
     inline void setInitialized(bool tf) { initialized = tf; }
     inline bool queryTimeActivities() const { return timeActivities; }
     inline roxiemem::RoxieHeapFlags queryHeapFlags() const { return defaultRoxieMemHeapFlags; }
+    inline bool traceStartStop() const { return false; }
```
Comment (jakesmith) [inline thorlcr/graph/thgraph.hpp:1139]: I assume placeholder for new flag and doTrace test.

File: thorlcr/master/thmastermain.cpp [hunk 8]
```diff
@@ -633,6 +633,7 @@ int main( int argc, const char *argv[]  )
     NoQuickEditSection xxx;
     {
         globals.setown(loadConfiguration(thorDefaultConfigYaml, argv, "thor", "THOR", "thor.xml", nullptr, nullptr, false));
+        updateTraceFlags(loadTraceFlags(globals, thorTraceOptions, TraceFlags::Standard), true);
```
Comment (jakesmith) [inline thorlcr/master/thmastermain.cpp:636]: maybe move to where logging is initialize (around line 700)

**Review by jakesmith — COMMENTED** [review]
@richardkchapman - looks sensible to me.

My only concern is that some the tracing has regularly been used (scanning all worker logs) to pinpoint where/why one/some workers are in a different state than others, removing it altogether will make diagnosis more difficult on those occasions.
If we continue to output some of this logigng to the circular local buffer (i.e. not to backend logging), that would still allow deeper investigations into stalls etc. where this tracing is helpful,

**Comment by ghalliday** [comment]
@richardkchapman @jakesmith 
I am going to close this for the moment.  I think the plan for Thor is to
a) disable all debug tracing to stdout.
b) Dump the post-mortem circular buffer whenever there is any job failure  
c) Add the option to dump the post-mortem circular buffers when a job is running.  (call it a snapshot in this case? all cases?)
d) Ensure multiple sets of log snapshots can be associated with a single workunit.
e) Ensure that there is a clean interface to extract them from eclwatch.

**Pull request closed without merging**